### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by Teollan**
[DEMO LINK](https://Teollan.github.io/react_carousel/)

### CODE REVIEW

**Review by AndriiZakharenko — APPROVED** [review]
good job!

File: src/App.tsx [hunk 1]
```diff
@@ -20,17 +25,87 @@ class App extends React.Component<{}, State> {
...
-        {/* eslint-disable-next-line */}
-        <h1>Carousel with {images.length} images</h1>
+        <h1 data-cy="title">Carousel with {this.state.images.length} images</h1>
+
+        <>
+          <label htmlFor="itemId">Item width: </label>
+          <input
+            onChange={event =>
+              this.setState({ itemWidth: +event.target.value })
+            }
```
Comment (BudnikOleksii) [inline src/App.tsx:46]: Create handlers instead of passing anonim functions to `onChange`

File: src/App.tsx [hunk 2]
```diff
@@ -20,17 +25,87 @@ class App extends React.Component<{}, State> {
...
+            type="number"
+            name="itemId"
+            id="itemId"
+            defaultValue={130}
```
Comment (BudnikOleksii) [inline src/App.tsx:50]: Why do you use `defaultValue`? The value should depend on the state, and these values are magic values, create a constant

File: src/components/Carousel.tsx [hunk 3]
```diff
@@ -1,26 +1,142 @@
...
+  animationDuration?: number;
+  infinite?: boolean;
+};
+
+const Carousel: React.FC<Props> = ({
+  images,
+  step = 3,
+  frameSize = 3,
+  itemWidth = 130,
+  animationDuration = 1000,
```
Comment (BudnikOleksii) [inline src/components/Carousel.tsx:18]: That's what I meant, you use these value in 3 places, create a constant so it will be the same in any case

File: src/components/Carousel.tsx [hunk 4]
```diff
@@ -1,26 +1,142 @@
...
+  infinite = false,
+}) => {
+  const [index, setIndex] = useState(0);
```
Comment (BudnikOleksii) [inline src/components/Carousel.tsx:21]: What index is it?

File: src/components/Carousel.tsx [hunk 5]
```diff
@@ -1,26 +1,142 @@
...
+  const [isPrevPossible, setIsPrevPossible] = useState(infinite || index > 0);
+  const [isNextPossible, setIsNextPossible] = useState(
+    infinite || index < images.length - frameSize,
+  );
```
Comment (BudnikOleksii) [inline src/components/Carousel.tsx:25]: These values are dynamic, don't store them in the state, just create variables and do calculations

File: src/components/Carousel.tsx [hunk 6]
```diff
@@ -1,26 +1,142 @@
...
+
+  const containerStyle = { width: `${containerWidth}px` };
+  const listStyle = {
+    left: `-${offset}px`,
+    transitionDuration: `${animationDuration}ms`,
+  };
+
+  const imageHiddenStyle = {
+    visibility: 'hidden',
+    transitionDelay: `${animationDuration}ms`, // creates an illusion of smooth motion without angering the cypress gods
```
Comment (BudnikOleksii) [inline src/components/Carousel.tsx:38]: Remove all comments

File: src/components/Carousel.tsx [hunk 7]
```diff
@@ -1,26 +1,142 @@
...
+        return;
+      }
+
+      setIsPrevPossible(newIndex > 0);
+      setIsNextPossible(newIndex < images.length - frameSize);
+    },
+    [frameSize, images.length, infinite],
+  );
+
+  const goToNextIndex = useCallback(() => {
```
Comment (BudnikOleksii) [inline src/components/Carousel.tsx:53]: It doesn't make sense to wrap these functions with `useCallback` as you don't pass them to other components

File: src/components/Carousel.tsx [hunk 8]
```diff
@@ -1,26 +1,142 @@
...
+    setIndex(oldIndex => {
+      let newIndex = oldIndex + step;
+
+      if (newIndex >= images.length - (frameSize - step)) {
+        newIndex = 0;
+      } else if (newIndex > images.length - frameSize) {
+        newIndex = images.length - frameSize;
+      }
+
+      updateMovement(newIndex);
```
Comment (BudnikOleksii) [inline src/components/Carousel.tsx:63]: I don't get what the point of these functions, but at least in the `setState` calls it's not a good idea to call other side functions

**Review by BudnikOleksii — CHANGES_REQUESTED** [review]
Good work 👍 
Left some comments, check them. Also, it would be great to add some validations, as now user can set any values and they can be invalid
image

File: src/components/Carousel.tsx (hunk 5 above)
Comment (Teollan) [reply src/components/Carousel.tsx:25]: Yep, you are right. I don't know what i was thinking when I wrote this 😅

File: src/components/Carousel.tsx (hunk 8 above)
Comment (Teollan) [reply src/components/Carousel.tsx:63]: The functions are supposed to calculate a correct index in some edge cases, like when `step !== frameSize` or when number of pictures isn't perfectly divisible by `frameSize`. And they work both for normal and infinite `Carousel`.

TLDR; it's a mess, but it's functional)

**Review by BudnikOleksii — APPROVED** [review]
👍
